core/consensus/aura/slots/src/lib.rs

```diff
@@ -0,0 +1,255 @@
+// Copyright 2019 Parity Technologies (UK) Ltd.
+// This file is part of Substrate.
+
+// Substrate is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+extern crate parity_codec as codec;
+
+mod slots;
+
+pub use slots::{Slots, SlotInfo};
+
+use std::sync::{mpsc, Arc};
+use std::thread;
+use futures::prelude::*;
+use futures::{Future, IntoFuture, future::{self, Either}};
+use log::{warn, debug, info};
+use runtime_primitives::generic::BlockId;
+use runtime_primitives::traits::{ProvideRuntimeApi, Block};
+use consensus_common::SyncOracle;
+use inherents::{InherentData, InherentDataProviders};
+use aura_primitives::AuraApi;
+use client::ChainHead;
+use codec::Encode;
+
+/// A worker that should be invoked at every new slot.
+pub trait SlotWorker<B: Block> {
+	type OnSlot: IntoFuture<Item=(), Error=consensus_common::Error>;
```

Comment:
Why not just `Future`?

Reply:
Because it's more convenient -- both `Future` and `Result` implements `IntoFuture`, so if some workers don't need future, they can just set `OnSlot` to be `Result`.

Reply:
For that we have `future::FutureResult`.

Reply:
Exactly. https://docs.rs/futures/0.1.25/src/futures/future/mod.rs.html#1061-1069

Reply:
Yeah, I know that implementation. But you are right. I had some problems with that lately but it was probably something different. Sorry for the noise :(